File: DotOther/Managed/Source/GarbageCollector.cs
using System;
using System.Runtime.InteropServices;

namespace DotOther.Managed {

  internal static class GarbageCollector {
    [UnmanagedCallersOnly]
    internal static void CollectGarbage(Int32 generation, GCCollectionMode mode, NBool32 blocking, NBool32 compacting) {
      try { 
        if (generation < 0) {
          GC.Collect();
        } else {
          GC.Collect(generation, mode, blocking, compacting);
        }
      } catch (Exception e) {
        DotOtherHost.HandleException(e);
      }
    }

    [UnmanagedCallersOnly]
    internal static void WaitForPendingFinalizers() {
      try {
        GC.WaitForPendingFinalizers();
      } catch (Exception e) {
        DotOtherHost.HandleException(e);
      }
    }
  }

}

File: DotOther/Native/hosting/garbage_collector.cpp
/**
 * \file hosting/garbage_collector.cpp
 **/
#include "hosting/garbage_collector.hpp"

#include "hosting/interop_interface.hpp"

namespace dotother {

  void GarbageCollector::Collect() {
    Collect(-1, GCMode::DEFAULT, true, false);
  }

  void GarbageCollector::CompactCollect() {
    Collect(-1, GCMode::DEFAULT, true, true);
  }

  void GarbageCollector::Collect(int32_t context_id, GCMode mode, bool blocking, bool compacting) {
    Interop().collect_garbage(context_id, mode, blocking, compacting);
  }

  void GarbageCollector::WaitForPendingFinalizers(int32_t context_id) {
    Interop().wait_for_pending_finalizers();
  }

} // namespace dotother